### DISCUSSION TITLE
nushell n00b, looking for some help with external completers

### DISCUSSION THREAD

**Discussion by JeffMelton** [nushell/nushell#10866] [Q&A] [answered]
Hey, folks, I just started digging in to nushell a couple days ago, and I'm really enjoying it. I'm sure this is a PEBKAC error, but external completers aren't working as expected for me, and I'm looking for some input as to why that might be.

Basically, what's happening is that, no matter what I do, the only completion I get is for the first item (e.g., `git`). When I try to complete any subcommand or flag from _any_ binary, the only completions offered are file paths.

I'm a former fish shell user, so I have a ton of completions from there. And I installed `carapace`, thinking that experience might be better. I've been using `zoxide` for some time, so that's onboard as well. All of my config is based off the [cookbook](https://www.nushell.sh/cookbook/external_completers.html), with just a couple changes for things that either appear to be typos (e.g., `do $in $spans`) or seem not to have been updated after some change (e.g., `scope alias-completions`.

From `config.nu`
```
let carapace_completer = {|spans|
  carapace $spans.0 nushell $spans | from json
}

let fish_completer = {|spans|
    fish --command $'complete "--do-complete=($spans | str join " ")"'
    | $"value(char tab)description(char newline)" + $in
    | from tsv --flexible --no-infer
}

let zoxide_completer = {|spans|
    $spans | skip 1 | zoxide query -l $in | lines | where {|x| $x != $env.PWD}
}

let external_completer = {|spans|
    let expanded_alias = scope aliases
    | where name == $spans.0
    | get -i 0.expansion

    let spans = if $expanded_alias != null {
        $spans
        | skip 1
        | prepend ($expanded_alias | split row ' ')
    } else {
        $spans
    }

    match $spans.0 {
        nu => $fish_completer
        git => $carapace_completer
        z => $zoxide_completer
        zi => $zoxide_completer
        _ => $carapace_completer
    } | do in $spans
}

...

    completions: {
        case_sensitive: false
        quick: true
        partial: true
        algorithm: "prefix"
        external: {
            enable: true
            max_results: 100
            completer: { $external_completer }
        }
    }
```

Like I say, I feel like this is probably just a config error somewhere, but for the life of me I can't figure out which knob to twiddle to make it work right. TIA.

**Comment by 101313** [comment]
Hey, I found that carapace isn't a good completer to set as a default and relative lacking in contrast to fish.
I currently reside solely at fish and my configuration looks like this 

```nushell
export-env {
   let external_completer = {|tokens: list<string>|
         let expanded_alias = scope aliases | where name == $tokens.0 | get --ignore-errors expansion.0

         let tokens = if $expanded_alias != null  {
             $expanded_alias | split row " " | append ($tokens | skip 1)
         } else {
             $tokens
         }

         let cmd = $tokens.0 | str trim --left --char '^'
         let hasFish = fish --command $"find $fish_complete_path* -iname ($cmd).fish 2>/dev/null" | into string

         if ($hasFish | is-empty) {
           carapace $cmd nushell $tokens | from json | default []
         } else  {
         fish  -c $'complete "-kC ($tokens | str join " ")"' | from tsv -n | rename value description
         }
  }

   $env.config.completions = {
       case_sensitive: false
       quick: true
       partial: true
       algorithm: "prefix"
       external: {
         enable: true
         max_results: 100
  
          completer: $external_completer
       }
   }
}
```

While certainly not perfect, it might serve you as a good starting point or inspiration

**Reply by JeffMelton** [reply]
Hmm, that's not working as expected either. >.<

**Comment by 101313** [comment]
Hey, try replacing your match line with
```nushell
match $"($spans.0) " {
```

**Reply by JeffMelton** [reply]
No dice. ¯\_(ツ)_/¯ I'd love to get it working, but I feel like that's jumping in to the deep end of the pool. There's a lot I don't understand, and the docs are pretty thin still. Thanks for trying to help. Gonna close this and move on. :)

**Reply by 101313** [reply]
Is this what you're after?
https://asciinema.org/a/x42JWOvGbAmzqt7S1v4uQpaT8

My setup currently:
```
completer = {|spans|

    let fish_completer = {|spans|
      fish --command $'complete "--do-complete=($spans | str join " ")"'
      | $"value(char tab)description(char newline)" + $in
      | from tsv --flexible --no-infer
    }

    let expanded_alias = scope aliases
    | where name == $spans.0
    | get -i 0.expansion

    let spans = if $expanded_alias != null {
        $spans
        | skip 1
        | prepend ($expanded_alias | split row ' ')
    } else {
        $spans 
    }

    match $"($spans.0) " {
        _ => $fish_completer
    } | do $in $spans
 
}

```

**Reply by JeffMelton** [reply]
I guess I don't understand where that fits in context. How are you calling it? I'd expect to see a `let` in front of that block.

Maybe it's less a matter of my syntax being wrong than it is that I'm putting things in the wrong place or calling them incorrectly. I'm setting completers in config.nu.

**Comment by 101313** [comment, marked as answer]
Hey sorry, you're right, my answer can be confusing.

```completer = {|spans|```  is essentially the object that goes into $env.config.completions.external.completer.

You can either replace your external_completer with it, or assign it directly to your config like so:
```nushell
$env.config.completions = {
    case_sensitive: false
    quick: true
    partial: true
    algorithm: "prefix"
    external: {
      enable: true
      max_results: 100
      completer: {|spans|
        let fish_completer = {|spans|
          fish --command $'complete "--do-complete=($spans | str join " ")"'
          | $"value(char tab)description(char newline)" + $in
          | from tsv --flexible --no-infer
        }

        let expanded_alias = scope aliases
        | where name == $spans.0
        | get -i 0.expansion

        let spans = if $expanded_alias != null {
            $spans
            | skip 1
            | prepend ($expanded_alias | split row ' ')
        } else {
            $spans 
        }

        match $"($spans.0) " {
            _ => $fish_completer
        } | do $in $spans
      }
    }
}
```
Hope it worked out =)

**Reply by 101313** [reply]
By the way, I think just adding ```| insert 1 " "``` in here
```nushell
    } else {
        $spans 
    }
```

Will probably do the trick, without having to further modify any of your config.
Just make sure you use fish_completer as the default case inside the match clause and not carapace.

**Reply by JeffMelton** [reply]
Thank you! Adding `| insert 1 " "` didn't work, but putting that block in as the completer did. Killer. 'preciate it!
